swap/simulations_test.go

```diff
@@ -0,0 +1,779 @@
+		},
+		// and we also use a swap service
+		"swap": func(ctx *adapters.ServiceContext, bucket *sync.Map) (s node.Service, cleanup func(), err error) {
+			// every simulation node has an instance of a `testService`
+			ts := newTestService()
+			// balance is the interface for `NewAccounting`; it is a Swap
+			balance := params.swaps[params.count]
+			dir := params.dirs[params.count]
+			// every node is a different instance of a Swap and gets a different entry in the map
+			params.count++
+			// to create the accounting, we also need a `Price` instance
+			prices := &testPrices{}
+			// create the matrix of test prices
+			prices.newTestPriceMatrix()
+			ts.spec = newTestSpec()
+			// create the accounting instance and assign to the spec
+			ts.spec.Hook = protocols.NewAccounting(balance, prices)
+			ts.swap = balance
+			// deploy the accounting to the `SimulatedBackend`
+			err = testDeploy(context.Background(), balance.backend, balance)
+			if err != nil {
+				return nil, nil, err
+			}
+			// store the testService into the bucket
+			bucket.Store(bucketKeySwap, ts)
+
+			cleanup = func() {
+				ts.swap.store.Close()
+				os.RemoveAll(dir)
+			}
+
+			return ts, cleanup, nil
+		},
+	}
+	return simServiceMap
+}
+
+// newSharedBackendSwaps pre-loads each simulated node account with "funds"
+// so that later in the simulation all operations have sufficient gas
+func newSharedBackendSwaps(nodeCount int) (*swapSimulationParams, error) {
+	params := &swapSimulationParams{
+		swaps:       make(map[int]*Swap),
+		dirs:        make(map[int]string),
+		maxMsgPrice: 10000,
+		minMsgPrice: 100,
+		nodeCount:   nodeCount,
+	}
+	keys := make(map[int]*ecdsa.PrivateKey)
+	addrs := make(map[int]common.Address)
+	alloc := core.GenesisAlloc{}
+	stores := make(map[int]*state.DBStore)
+
+	// for each node, generate keys, a GenesisAccount and a state store
+	for i := 0; i < nodeCount; i++ {
+		key, err := crypto.GenerateKey()
+		if err != nil {
+			return nil, err
+		}
+		keys[i] = key
+		addrs[i] = crypto.PubkeyToAddress(key.PublicKey)
+		alloc[addrs[i]] = core.GenesisAccount{Balance: big.NewInt(10000000000)}
+		dir, err := ioutil.TempDir("", fmt.Sprintf("swap_test_store_%x", addrs[i].Hex()))
+		if err != nil {
+			return nil, err
+		}
+		stateStore, err2 := state.NewDBStore(dir)
+		if err2 != nil {
+			return nil, err
+		}
+		params.dirs[i] = dir
+		stores[i] = stateStore
+	}
+	// then create the single SimulatedBackend
+	gasLimit := uint64(8000000000)
+	defaultBackend := backends.NewSimulatedBackend(alloc, gasLimit)
+	testBackend := &swapTestBackend{SimulatedBackend: defaultBackend}
+	// finally, create all Swap instances for each node, which share the same backend
+	for i := 0; i < nodeCount; i++ {
+		params.swaps[i] = New(stores[i], keys[i], testBackend)
+	}
+
+	params.backend = testBackend
+	return params, nil
+}
+
+// TestPingPongChequeSimulation just launches two nodes and sends each other
```

Comment:
strictly speaking, in these simulations no message sending (to trigger accounting) should be done since that is tested in accounting pkg. instead just a process that periodically calls `Add` on each connection. 
Just run the swap protocol and see if the cheques are sent

Reply:
We want to test actual `Swap` code (package `swap`) here, not `p2p/accounting`. This specific simulation for example tests that without triggering cheques, the `Swap` code runs correctly.

Reply:
I think that's what @zelig actually means here. By triggering `Swap` through sending messages we run (and therefore also test) the `p2p/accounting` code which calls our `Add`. By calling `Add` on both peers directly (with signs reversed) on every message instead we would test only the actual `Swap` code without needing the fake protocol in the middle.

Reply:
I insist this is a first level of integration tests which do make sense. At no other point there is an end-to-end simulation test of sending messages and accounting done in one go.